src/cmd/services/m3coordinator/ingest/ingest.go

```diff
@@ -0,0 +1,195 @@
+	RetryOptions      retry.Options
+	InstrumentOptions instrument.Options
+}
+
+type ingestMetrics struct {
+	ingestError   tally.Counter
+	ingestSuccess tally.Counter
+}
+
+func newIngestMetrics(scope tally.Scope) ingestMetrics {
+	return ingestMetrics{
+		ingestError:   scope.Counter("ingest-error"),
+		ingestSuccess: scope.Counter("ingest-success"),
+	}
+}
+
+// Ingester ingests metrics with a worker pool.
+type Ingester struct {
+	workers xsync.PooledWorkerPool
+	p       pool.ObjectPool
+}
+
+// NewIngester creates an ingester.
+func NewIngester(
+	opts *Options,
+) *Ingester {
+	retrier := retry.NewRetrier(opts.RetryOptions)
+	m := newIngestMetrics(opts.InstrumentOptions.MetricsScope())
+	p := pool.NewObjectPool(opts.PoolOptions)
+	p.Init(
+		func() interface{} {
+			// NB: we don't need a pool for the tag decoder since the ops are
+			// pooled, but currently this is the only way to get tag decoder.
+			tagDecoder := opts.TagDecoderPool.Get()
+			op := ingestOp{
+				s:  opts.Appender,
+				r:  retrier,
+				it: serialize.NewMetricTagsIterator(tagDecoder, nil),
+				p:  p,
+				m:  m,
+			}
+			op.attemptFn = op.attempt
+			op.ingestFn = op.ingest
+			return &op
+		},
+	)
+	return &Ingester{
+		workers: opts.Workers,
+		p:       p,
+	}
+}
+
+// Ingest ingests a metric asynchronously with callback.
+func (i *Ingester) Ingest(
+	id []byte,
+	metricTimeNanos int64,
+	value float64,
+	sp policy.StoragePolicy,
+	callback *m3msg.RefCountedCallback,
+) {
+	op := i.p.Get().(*ingestOp)
+	op.id = id
+	op.metricTimeNanos = metricTimeNanos
+	op.value = value
+	op.sp = sp
+	op.callback = callback
+	i.workers.Go(op.ingestFn)
+}
+
+type ingestOp struct {
+	s         storage.Appender
+	r         retry.Retrier
+	it        id.SortedTagIterator
+	p         pool.ObjectPool
+	m         ingestMetrics
+	attemptFn retry.Fn
+	ingestFn  func()
+
+	id              []byte
+	metricTimeNanos int64
+	value           float64
+	sp              policy.StoragePolicy
+	callback        *m3msg.RefCountedCallback
+	q               storage.WriteQuery
+}
+
+func (op *ingestOp) ingest() {
+	if err := op.resetWriteQuery(); err != nil {
+		op.m.ingestError.Inc(1)
+		op.callback.Callback(m3msg.OnRetriableError)
+		op.p.Put(op)
+		return
+	}
+	if err := op.r.Attempt(op.attemptFn); err != nil {
+		if xerrors.IsNonRetryableError(err) {
+			op.callback.Callback(m3msg.OnNonRetriableError)
+		} else {
+			op.callback.Callback(m3msg.OnRetriableError)
+		}
+		op.m.ingestError.Inc(1)
+		op.p.Put(op)
+		return
+	}
+	op.m.ingestSuccess.Inc(1)
+	op.callback.Callback(m3msg.OnSuccess)
+	op.p.Put(op)
+}
+
+func (op *ingestOp) attempt() error {
+	return op.s.Write(
+		// NB: No timeout is needed for this as the m3db client has a timeout
+		// configured to it.
+		context.Background(),
+		&op.q,
+	)
+}
+
+func (op *ingestOp) resetWriteQuery() error {
+	if err := op.resetTags(); err != nil {
+		return err
+	}
+	op.resetDataPoints()
+	op.q.Raw = string(op.id)
+	op.q.Unit = op.sp.Resolution().Precision
```

Comment:
I just checked how MaxUnit gets resolved for say a storage policy of `1m:30d` and it seems like it's going to set Unit of `xtime.Minute` which won't be compatible with M3DB (which at most supports a unit/precision of time.Second).

Reply:
I'll just use `xtime.Millisecond` then, seems that's the default value used in several places in query

Reply:
Updating the usage in flush_handler.go as well